src/topology/local_homeomorph.lean

```diff
@@ -0,0 +1,479 @@
+    apply continuous_on.comp e.continuous_inv_fun e'.continuous_inv_fun,
+    rw [e'.to_local_equiv.inv_image_target_eq_source, h],
+  end,
+  ..local_equiv.trans' e.to_local_equiv e'.to_local_equiv h }
+
+/-- Composing two local homeomorphisms, by restricting to the maximal domain where their
+composition is well defined. -/
+protected def trans : local_homeomorph α γ :=
+  local_homeomorph.trans' (e.symm.restr_open e'.source e'.open_source).symm
+    (e'.restr_open e.target e.open_target) (by simp [inter_comm])
+
+@[simp] lemma trans_to_local_equiv :
+  (e.trans e').to_local_equiv = e.to_local_equiv.trans e'.to_local_equiv := rfl
+
+@[simp] lemma trans_to_fun : (e.trans e').to_fun = e'.to_fun ∘ e.to_fun := rfl
+@[simp] lemma trans_inv_fun : (e.trans e').inv_fun = e.inv_fun ∘ e'.inv_fun := rfl
+
+lemma trans_symm_eq_symm_trans_symm : (e.trans e').symm = e'.symm.trans e.symm :=
+by cases e; cases e'; refl
+
+/- This could be considered as a simp lemma, but there are many situations where it makes something
+simple into something more complicated. -/
+lemma trans_source : (e.trans e').source = e.source ∩ e.to_fun ⁻¹' e'.source :=
+local_equiv.trans_source e.to_local_equiv e'.to_local_equiv
+
+lemma trans_source' : (e.trans e').source = e.source ∩ e.to_fun ⁻¹' (e.target ∩ e'.source) :=
+local_equiv.trans_source' e.to_local_equiv e'.to_local_equiv
+
+lemma trans_source'' : (e.trans e').source = e.inv_fun '' (e.target ∩ e'.source) :=
+local_equiv.trans_source'' e.to_local_equiv e'.to_local_equiv
+
+lemma image_trans_source : e.to_fun '' (e.trans e').source = e.target ∩ e'.source :=
+local_equiv.image_trans_source e.to_local_equiv e'.to_local_equiv
+
+lemma trans_target : (e.trans e').target = e'.target ∩ e'.inv_fun ⁻¹' e.target := rfl
+
+lemma trans_target' : (e.trans e').target = e'.target ∩ e'.inv_fun ⁻¹' (e'.source ∩ e.target) :=
+trans_source' e'.symm e.symm
+
+lemma trans_target'' : (e.trans e').target = e'.to_fun '' (e'.source ∩ e.target) :=
+trans_source'' e'.symm e.symm
+
+lemma inv_image_trans_target : e'.inv_fun '' (e.trans e').target = e'.source ∩ e.target :=
+image_trans_source e'.symm e.symm
+
+lemma trans_assoc (e'' : local_homeomorph γ δ) :
+  (e.trans e').trans e'' = e.trans (e'.trans e'') :=
+eq_of_local_equiv_eq $ local_equiv.trans_assoc e.to_local_equiv e'.to_local_equiv e''.to_local_equiv
+
+@[simp] lemma trans_refl : e.trans (local_homeomorph.refl β) = e :=
+eq_of_local_equiv_eq $ local_equiv.trans_refl e.to_local_equiv
+
+@[simp] lemma refl_trans : (local_homeomorph.refl α).trans e = e :=
+eq_of_local_equiv_eq $ local_equiv.refl_trans e.to_local_equiv
+
+lemma trans_of_set {s : set β} (hs : is_open s) :
+  e.trans (of_set s hs) = e.restr (e.to_fun ⁻¹' s) :=
+local_homeomorph.ext _ _ (λx, rfl) (λx, rfl) $
+  by simp [local_equiv.trans_source, (e.preimage_interior _).symm, interior_eq_of_open hs]
+
+lemma trans_of_set' {s : set β} (hs : is_open s) :
+  e.trans (of_set s hs) = e.restr (e.source ∩ e.to_fun ⁻¹' s) :=
```

Comment:
Would it make sense to separate out the lemma `e.restr (e.source ∩ t) = e.restr t`?

Reply:
Done